### PR TITLE
Parse k8s set with mapstructure

### PR DESCRIPTION
**Pull request by neoaggelos**
### Summary

Use `mapstructure` to parse `k8s set` command arguments

### Changes

- Allows parsing string slice inputs as `val1,val2` or YAML (or JSON) formatted strings
- Add unit tests for `k8s set` arguments

### Notes

- If `mapstructure.NewDecoder` fails we panic, as all errors are programming bugs (see implementation of `NewDecoder`)
- We check for the list of known keys such that we present a meaningful user error, rather than a mapstructure error that the field is unknown.

### CODE REVIEW

File: src/k8s/cmd/k8s/k8s_set.go [hunk 1]
```diff
@@ -66,119 +65,69 @@ func newSetCmd(env cmdutil.ExecutionEnvironment) *cobra.Command {
...
+		return fmt.Errorf("unknown option key %q", key)
+	}
+
+	if err := decoder.Decode(toRecursiveMap(key, value)); err != nil {
+		return fmt.Errorf("invalid option %q: %w", arg, err)
 	}
 	return nil
 }
+
+func toRecursiveMap(key, value string) map[string]any {
```
Comment (bschimke95) [inline src/k8s/cmd/k8s/k8s_set.go:127]: nit: does it make sense to move it to `util/mapstructure.go`?

File: src/k8s/pkg/utils/mapstructure.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	"reflect"
+	"strings"
+	"unicode"
+
+	"github.com/mitchellh/mapstructure"
+	"gopkg.in/yaml.v2"
+)
+
+// YAMLToStringSliceHookFunc returns a mapstructure.DecodeHookFunc that converts string to []string by parsing YAML.
+func YAMLToStringSliceHookFunc(f reflect.Kind, t reflect.Kind, data interface{}) (interface{}, error) {
```
Comment (bschimke95) [inline src/k8s/pkg/utils/mapstructure.go:13]: nit: `any` instead of `interface{}`

**Review by bschimke95 — APPROVED** [review]
LGTM

File: src/k8s/cmd/k8s/k8s_set.go [hunk 3]
```diff
@@ -66,119 +65,69 @@ func newSetCmd(env cmdutil.ExecutionEnvironment) *cobra.Command {
...
-	case "metrics-server.enabled":
-		v, err := strconv.ParseBool(value)
-		if err != nil {
-			return fmt.Errorf("invalid boolean value for metrics-server.enabled: %w", err)
-		}
-		config.MetricsServer.Enabled = &v
-	default:
-		return fmt.Errorf("unknown config key %q", key)
+	if _, ok := knownSetKeys[key]; !ok {
+		return fmt.Errorf("unknown option key %q", key)
```
Comment (louiseschmidtgen) [inline src/k8s/cmd/k8s/k8s_set.go:118]: Should we add "consult `k8s get` command to see configurable options to the error message?

File: src/k8s/cmd/k8s/k8s_set.go (hunk 1 above)
Comment (neoaggelos) [reply src/k8s/cmd/k8s/k8s_set.go:127]: let's keep it in mind when we need in a second place in the code base.

File: src/k8s/pkg/utils/mapstructure.go (hunk 2 above)
Comment (neoaggelos) [reply src/k8s/pkg/utils/mapstructure.go:13]: going by https://github.com/mitchellh/mapstructure/blob/main/mapstructure.go#L185-L197

**Pull request merged**